apis/topology/v1beta1/topology_types.go

```diff
@@ -0,0 +1,136 @@
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/affinity"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"context"
+)
+
+// TopologySpec defines the desired state of Topology
+type TopologySpec struct {
+	// +kubebuilder:validation:Optional
+	// APITopologySpreadConstraint exposes topologySpreadConstraint that are
+	// applied to the StatefulSet
+	TopologySpreadConstraint *[]corev1.TopologySpreadConstraint `json:"topologySpreadConstraint,omitempty"`
+
+	// APIAffinity exposes PodAffinity and PodAntiaffinity overrides that are applied
+	// to the StatefulSet
+	// +optional
+	APIAffinity *affinity.Overrides `json:",inline"`
+
+	//TODO: We could add NodeSelector here as it belongs to the same APIGroup
+}
+
+// TopologyStatus defines the observed state of Topology
+type TopologyStatus struct {
+	// Conditions
+	Conditions condition.Conditions `json:"conditions,omitempty" optional:"true"`
+
+	// Hash of the topology configuration
+	Hash string `json:"hash,omitempty"`
+
+	// ObservedGeneration - the most recent generation observed for this
+	// service. If the observed generation is less than the spec generation,
+	// then the controller has not processed the latest changes injected by
+	// the opentack-operator in the top-level CR (e.g. the ContainerImage)
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// Topology is the Schema for the topologies API
+type Topology struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TopologySpec   `json:"spec,omitempty"`
+	Status TopologyStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// TopologyList contains a list of Topology
+type TopologyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Topology `json:"items"`
+}
+
+func init() {
+	SchemeBuilder.Register(&Topology{}, &TopologyList{})
+}
+
+// IsReady returns true if TopologySpreadConstraint reconciled successfully
+func (instance Topology) IsReady() bool {
+	return instance.Status.Conditions.IsTrue(condition.ReadyCondition)
+}
+
+// GetTopologyWithName - a function exposed to the service operators
+// that need to retrieve the referenced topology by name
+func GetTopologyWithName(
+	ctx context.Context,
+	h *helper.Helper,
+	name string,
+	namespace string,
+) (*Topology, string, error) {
+
+	topology := &Topology{}
+	err := h.GetClient().Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, topology)
+	if err != nil {
+		return topology, "", err
+	}
+	return topology, topology.Status.Hash, nil
+}
+
+
+// AddFinalizer -
+func (t *Topology) AddFinalizer(
```

Comment:
controllerutil.AddFinalizer already implement the logic to only add a finalizer if it is not present.

Reply:
Regarding this, my idea is that the same CR that stores this configuration can be used by multiple CRs. When that happens, the caller put a finalizer on it. I understand `controllerutil` has the logic of checking if the finalizer is present, but does that work even when an external entity references this CR? I assume you have to fetch it, put the finalizer and update it explicitly.
I updated this code to use `controllerutils` and get rid of all that custom writing, but I'm wondering if there's a way to not implement this function at all.

Reply:
Done, I think this logic is not needed here.